internal/statesync/reactor.go

```diff
@@ -979,43 +995,36 @@ func (r *Reactor) processPeerUpdate(ctx context.Context, peerUpdate p2p.PeerUpda
 			peerUpdate.Channels.Contains(ParamsChannel) {
 
 			r.peers.Append(peerUpdate.NodeID)
-
 		} else {
 			r.logger.Error("could not use peer for statesync", "peer", peerUpdate.NodeID)
 		}
-
-	case p2p.PeerStatusDown:
-		r.peers.Remove(peerUpdate.NodeID)
-	}
-
-	r.mtx.Lock()
-	defer r.mtx.Unlock()
-
-	if r.syncer == nil {
-		return
-	}
-
-	switch peerUpdate.Status {
-	case p2p.PeerStatusUp:
-
 		newProvider := NewBlockProvider(peerUpdate.NodeID, r.chainID, r.dispatcher)
 
-		r.providers[peerUpdate.NodeID] = newProvider
-		err := r.syncer.AddPeer(ctx, peerUpdate.NodeID)
-		if err != nil {
-			r.logger.Error("error adding peer to syncer", "error", err)
-			return
+		stateProvider := r.getStateProvider()
+		if stateProvider != nil {
+			if sp, ok := stateProvider.(*stateProviderP2P); ok {
+				// we do this in a separate routine to not block whilst waiting for the light client to finish
+				// whatever call it's currently executing
+				go sp.addProvider(newProvider)
+			}
 		}
-		if sp, ok := r.stateProvider.(*stateProviderP2P); ok {
-			// we do this in a separate routine to not block whilst waiting for the light client to finish
-			// whatever call it's currently executing
-			go sp.addProvider(newProvider)
+
+		syncer := r.getSyncer()
+		if syncer != nil {
+			if err := syncer.AddPeer(ctx, peerUpdate.NodeID); err != nil {
+				r.logger.Error("error adding peer to syncer", "error", err)
+				return
+			}
 		}
 
 	case p2p.PeerStatusDown:
-		delete(r.providers, peerUpdate.NodeID)
-		r.syncer.RemovePeer(peerUpdate.NodeID)
+		r.peers.Remove(peerUpdate.NodeID)
```

Comment:
WDYT to create a separate method like `removePeer` and `addPeer` for such kind of logic?